dao/authentication_dao.go

```diff
@@ -315,15 +315,54 @@ func (a *authenticationDaoImpl) Delete(uid string) (*m.Authentication, error) {
 	for _, key := range keys {
 		if strings.HasSuffix(key, uid) {
 			path := fmt.Sprintf("secret/metadata/%d/%s", *a.TenantID, key)
-			sec, err := Vault.Delete(path)
+			// fetch the key (for returning)
+			sec, _ := a.getKey(key)
+			// ...then delete it
+			_, err := Vault.Delete(path)
 
-			return authFromVault(sec), err
+			return sec, err
 		}
 	}
 
 	return nil, util.NewErrNotFound("authentication")
 }
 
+// Deletes all authentications for a parent resource. Sources are handled
+// differently than other parent resources mostly due to the cascade effect from
+// a source delete
+func (a *authenticationDaoImpl) Cleanup(resourceType string, resourceID int64) error {
+	resourceType = util.Capitalize(resourceType)
+
+	auths, _, err := a.List(100, 0, []util.Filter{})
+	if err != nil {
+		return err
+	}
+
+	for i := range auths {
+		if resourceType == "Source" {
+			if auths[i].SourceID != resourceID {
+				continue
+			}
+
+			_, err := a.Delete(auths[i].ID)
+			if err != nil {
+				return err
+			}
+		} else {
+			if auths[i].ResourceID != resourceID || auths[i].ResourceType != resourceType {
+				continue
+			}
+
+			_, err := a.Delete(auths[i].ID)
+			if err != nil {
+				return err
+			}
+		}
+	}
```

Comment:
I would suggest using `for _, auth := range auths` to avoid using the `auths[i]` statement inside the loop. I say this because I don't see us doing anything special with the index.